lib/spack/spack/cmd/env.py

```diff
@@ -472,6 +473,32 @@ def env_remove(args):
         tty.msg(f"Successfully removed environment '{bad_env_name}'")
 
 
+#
+# env rename
+#
+def env_rename_setup_parser(subparser):
+    """rename an existing environment"""
+    subparser.add_argument(
+        "mv_old_name", metavar="old_name", help="name of existing managed environment to rename"
+    )
+    subparser.add_argument("mv_new_name", metavar="new_name", help="new name for environment")
+
+
+def env_rename(args):
+    """Rename a *named* environment.
+
+    This renames a named environment managed by Spack.
+    """
+    env = ev.read(args.mv_old_name)
+    if args.mv_old_name == args.mv_new_name:
+        tty.warn("The old and new environment names are the same (no renaming done)")
+        return
+
+    new_path = fs.join_path(ev.environment.env_root_path(), args.mv_new_name)
```

Comment:
You could do a conditional check her to allow for anonymous environments too.
```suggestion
    if ev.is_env_dir(args.mv_old_name): # probably already get this check with ev.read actually
        if ev.exists(args.mv_old_name):
            new_path = fs.join_path(ev.environment.env_root_path(), args.mv_new_name)
        else:
            new_path = args.mv_new_name
    else:
        tty.msg("The renaming operation is not on a valid spack environment")
```
Some thoughts on how to expand to anonymous environments.

Reply:
If you do that, it should probably also check whether the new name is an absolute path or a name and do different things accordingly.

Reply:
I've decided to add a `-d` flag that indicates that the specified arguments correspond to paths and not names.  I hope to update the PR soon with this flag.

Reply:
See the updated implementation.